**#0 Importy**

In [170]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

**#1 Wczytanie danych**

In [171]:
data = load_iris()
print(data.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


In [172]:
print(data['data'][:5])  # Pierwsze 5 próbek
print(data['feature_names'])  # Nazwy cech
print(data['target'][:5])  # Pierwsze 5 etykiet klas
print(data['target_names'])  # Nazwy gatunków

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
[0 0 0 0 0]
['setosa' 'versicolor' 'virginica']


In [173]:
X = data['data'] 
y = data['target']

In [174]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [175]:
print(X_train.shape, X_test.shape)

(120, 4) (30, 4)


In [176]:
# Standaryzacja
scaler = StandardScaler()
X_train_standardized = scaler.fit_transform(X_train)
X_test_standardized = scaler.transform(X_test)

**#2 Deklaracja modelu**

In [177]:
model = Sequential([
    Dense(64, activation='relu', input_dim=4),  # 4 cechy wejściowe
    Dense(32, activation='relu'),
    Dense(3, activation='softmax')   # 3 klasy → softmax
])

C:\Users\Mateusz\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


**#3 Kompilacja modelu**

In [178]:
adam_opt = tf.keras.optimizers.Adam(learning_rate=0.001)

In [179]:
model.compile(
    optimizer=adam_opt,
    loss="sparse_categorical_crossentropy",  # Bo etykiety są liczbami (0,1,2)
    metrics=["accuracy"]
)

**#4 Podgląd modelu**

In [180]:
model.summary()

Model: "sequential_14"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_42 (Dense)                     │ (None, 64)                  │             320 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_43 (Dense)                     │ (None, 32)                  │           2,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_44 (Dense)                     │ (None, 3)                   │              99 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 2,499 (9.76 KB)

 Trainable params: 2,499 (9.76 KB)

 Non-trainable params: 0 (0.00 B)

**#5 Trenowanie modelu**

In [181]:
history = model.fit(
    X_train_standardized, 
    y_train, 
    epochs=50, 
    validation_split=0.2  # 20% danych treningowych na walidację
)

Epoch 1/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 2s 144ms/step - accuracy: 0.2643 - loss: 1.1522 - val_accuracy: 0.3750 - val_loss: 1.0653
Epoch 2/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - accuracy: 0.3581 - loss: 1.0734 - val_accuracy: 0.5417 - val_loss: 1.0052
Epoch 3/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - accuracy: 0.5391 - loss: 1.0138 - val_accuracy: 0.7083 - val_loss: 0.9519
Epoch 4/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - accuracy: 0.6927 - loss: 0.9474 - val_accuracy: 0.7917 - val_loss: 0.9032
Epoch 5/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - accuracy: 0.7669 - loss: 0.9063 - val_accuracy: 0.8333 - val_loss: 0.8599
Epoch 6/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 53ms/step - accuracy: 0.8581 - loss: 0.8455 - val_accuracy: 0.8333 - val_loss: 0.8192
Epoch 7/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - accuracy: 0.8438 - loss: 0.8135 - val_accuracy: 0.8333 - val_loss: 0.7819
Epoch 8/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - accuracy: 0.8307 - loss: 0.7679 - val_accuracy: 0.8750 - val_loss: 0.7459

**#6 Ocena modelu**

In [182]:
test_loss, test_acc = model.evaluate(X_test_standardized, y_test)
print(f"Dokładność na zbiorze testowym: {test_acc:.2f}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 65ms/step - accuracy: 0.8667 - loss: 0.2359
Dokładność na zbiorze testowym: 0.87


**#7 Podsumowanie**

Podsumowując, dokładność 87 % jest dobrą wartością dla klasyfikacji Iris, co oznacza, że model radzi sobie dobrze w rozróżnianiu gatunków kwiatów na podstawie ich cech.